MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+    MathLib::Point3d const& end_pnt(segment.getEndPoint());
+
+    for (auto const elem : surface_elements) {
+        // compute element-segment-intersections (2d in x-y-plane)
+        std::vector<MathLib::Point3d> element_intersections(
+            computeElementSegmentIntersections(*elem, segment));
+        if (element_intersections.empty())
+            continue;
+
+        BaseLib::makeVectorUnique(element_intersections);
+
+        std::vector<GeoLib::LineSegment> sub_seg_elem(
+            createSubSegmentsForElement(element_intersections, beg_elem,
+                                        end_elem, beg_pnt, end_pnt, elem));
+        sub_segments.insert(sub_segments.end(), sub_seg_elem.begin(),
+                            sub_seg_elem.end());
+    }
+
+    // beg_elem == nullptr means there isn't any element corresponding to the
+    // beg_pnt and as a consequence the above algorithm doesn't insert a sub
+    // segment
+    if (beg_elem == nullptr)
+    {
+        auto min_dist_segment = std::min_element(
+            sub_segments.begin(), sub_segments.end(),
+            [&beg_pnt](GeoLib::LineSegment const& seg0,
+                       GeoLib::LineSegment const& seg1)
+            {
+                // min dist for segment 0
+                const double d0(
+                    std::min(MathLib::sqrDist(beg_pnt, seg0.getBeginPoint()),
+                             MathLib::sqrDist(beg_pnt, seg0.getEndPoint())));
+                // min dist for segment 1
+                const double d1(
+                    std::min(MathLib::sqrDist(beg_pnt, seg1.getBeginPoint()),
+                             MathLib::sqrDist(beg_pnt, seg1.getEndPoint())));
+                return d0 < d1;
+            });
+        GeoLib::Point * pnt{
+            MathLib::sqrDist(beg_pnt, min_dist_segment->getBeginPoint()) <
+                    MathLib::sqrDist(beg_pnt, min_dist_segment->getEndPoint())
+                ? new GeoLib::Point{min_dist_segment->getBeginPoint()}
+                : new GeoLib::Point{min_dist_segment->getEndPoint()}};
+        sub_segments.emplace_back(
+            GeoLib::LineSegment{new GeoLib::Point{beg_pnt, 0}, pnt, true});
+    }
+    // sort all sub segments for the given segment (beg_pnt, end_pnt)
+    GeoLib::sortSegments(beg_pnt, sub_segments);
+
+    sub_segments.erase(std::unique(sub_segments.begin(), sub_segments.end()),
+                       sub_segments.end());
+
+    return sub_segments;
+}
+
+static void mapPointOnSurfaceElement(MeshLib::Element const& elem,
+                                     MathLib::Point3d& q)
+{
+    // create plane equation: n*p = d
+    MathLib::Vector3 const& p(*(elem.getNode(0)));
+    MathLib::Vector3 const n(MeshLib::FaceRule::getSurfaceNormal(&elem));
+    if (n[2] == 0.0) { // vertical plane, z coordinate is arbitrary
+        q[2] = p[2];
+    } else {
+        double const d(MathLib::scalarProduct(n, p));
+        q[2] = (d - n[0]*q[0] - n[1]*q[1])/n[2];
```

Comment:
You only modify z coordinates. I.e., that's not an orthogonal projection of the point `q` onto the element. Is that what you want?